### ISSUE TITLE
Transform: RefId choices disappear when using "Filter by refId"

### ISSUE THREAD

**Issue by kylebrandt** [grafana/grafana#24061]
**What would you like to be added**:

At some point in the past, I could use Transform's Filter By RefId to select a response from Dashboard query that was not a time series. It seems now I only see series.

**Why is this needed**:

With expressions, I have something that queries a datasource for a bunch of series, gets the average of those series, and then multiplies each series by its corresponding average.

The average of each series, is not a series. I would like to be able to filter to those frame(s) to display a table.

Selected Dashboard Query:

![image](https://user-images.githubusercontent.com/1692624/80611781-1206dd80-8a09-11ea-9c4a-5f221f351ec0.png)

Filter by refid (missing "B"):

![image](https://user-images.githubusercontent.com/1692624/80611798-18955500-8a09-11ea-906e-124d65413172.png)

Edit: I think maybe B's results are getting included in C here?

**Comment by ryantxu** [comment]
I don't really understand the question... i tried reproducting filtering by refId, and it seems to work as expected

<details>
  <summary>panel.json</summary>
  <pre>
{
  "datasource": null,
  "aliasColors": {},
  "bars": false,
  "dashLength": 10,
  "dashes": false,
  "fieldConfig": {
    "defaults": {
      "custom": {}
    },
    "overrides": []
  },
  "fill": 1,
  "fillGradient": 0,
  "gridPos": {
    "h": 9,
    "w": 12,
    "x": 0,
    "y": 0
  },
  "hiddenSeries": false,
  "id": 2,
  "legend": {
    "avg": false,
    "current": false,
    "max": false,
    "min": false,
    "show": true,
    "total": false,
    "values": false
  },
  "lines": true,
  "linewidth": 1,
  "nullPointMode": "null",
  "options": {
    "dataLinks": []
  },
  "percentage": false,
  "pointradius": 2,
  "points": false,
  "renderer": "flot",
  "seriesOverrides": [],
  "spaceLength": 10,
  "stack": false,
  "steppedLine": false,
  "targets": [
    {
      "refId": "A",
      "scenarioId": "random_walk"
    },
    {
      "refId": "B",
      "scenarioId": "random_walk"
    },
    {
      "refId": "C",
      "scenarioId": "random_walk"
    }
  ],
  "thresholds": [],
  "timeFrom": null,
  "timeRegions": [],
  "timeShift": null,
  "title": "Panel Title",
  "tooltip": {
    "shared": true,
    "sort": 0,
    "value_type": "individual"
  },
  "transformations": [
    {
      "id": "filterByRefId",
      "options": {
        "include": "A|C"
      }
    }
  ],
  "type": "graph",
  "xaxis": {
    "buckets": null,
    "mode": "time",
    "name": null,
    "show": true,
    "values": []
  },
  "yaxes": [
    {
      "format": "short",
      "label": null,
      "logBase": 1,
      "max": null,
      "min": null,
      "show": true
    },
    {
      "format": "short",
      "label": null,
      "logBase": 1,
      "max": null,
      "min": null,
      "show": true
    }
  ],
  "yaxis": {
    "align": false,
    "alignLevel": null
  }
}
</pre>
</details>

Is the problem that it is not shown in the graph?  Can you try the table view and see if it is there?

**Comment by kylebrandt** [comment]
I can't seem to figure this out, some combination of:
 - Editing the query the dashboard query is on
 - Reselecting the dashboard query
 - Hidden queries (in this case no data returned)
 - Removing Filter by RefId, refresh the panel, and then adding it back

I keep seeing various behaviors, and am having a hard time pinning down the conditions.

**Comment by kylebrandt** [comment]
I think I'm starting to get it maybe. I think the confusion for me is that what I can filter by refId is what is currently displayed, the the refIds I have in my query. 

So if I have dashboard query referencing a panel that contains queries with refIds A, B, C:
 - In my first case, I had hidden B, which made it so B returned no results. So I didn't see an option to select B.

Given the same shared query but no longer hidden, my confusion came after selecting B:

1. The Query

![image](https://user-images.githubusercontent.com/1692624/80717822-52c52c00-8ac7-11ea-8573-4132a2a6cf81.png)


2. Select my B Field:

![image](https://user-images.githubusercontent.com/1692624/80717879-6cff0a00-8ac7-11ea-8538-0c5cfe51c780.png)

3. Go Back to dashboard...

4. Go back to edit panel, transform view (A and C) are now gone.

![image](https://user-images.githubusercontent.com/1692624/80718086-af284b80-8ac7-11ea-8856-36945e370c18.png)

From here, if I hit trash icon on Filter by refId, and try to add it back, I still only see B like above. If I hit "refresh", I still only see B. If I hit refresh, delete the transform, and add it back, now I can see A, B, and C as options again.

So in short I think the *options* for "Filter by RefId" should be based on the refIds in the query, not the data. Else there is a lot of refreshing (and I am currently working with a slow datasource) - and for me confusion about the changing options in the transform.

**Comment by torkelo** [comment]
> "Filter by RefId" should be based on the refIds in the query, not the data. 

This is not possible, transforms only have access to the query results. 

I tested this a bunch now and was unable to replicate any problem, so very strange. 

I created one panel1, that 3 queries return 1 series each, shown in table. 

Then a separate panel sharing the results from panel1) , can see query A, B, C in transforms. And this does not change no matter what I select in the table series dropdown. If I into panel1 and hide (using eye icon) query B it, of course, is no longer visible in the filter by refId in panel2. 

So not sure what is going on in your case, is query B only sometimes returning results or are you disabling the query in edit mode, then going to panel 2 and expect it to be still there? that would be hard given there are not results for the disabled query.

**Comment by kylebrandt** [comment]
> So not sure what is going on in your case, is query B only sometimes returning results or are you disabling the query in edit mode, then going to panel 2 and expect it to be still there? that would be hard given there are not results for the disabled query.

I am not in the case of the above screenshots. That is all with it enabled. Was able to replicate it again now, the base query is slow, so maybe that does it?

**Comment by torkelo** [comment]
Ah, that explains it, if one query is very late that refId won’t show (until the query completes)

**Comment by kylebrandt** [comment]
@torkelo https://youtu.be/9wtwlYFbnRk

**Comment by ryantxu** [comment]
I think I see where we need to dig in here -- the of optoins is populated when the component first loads, but does not detect any changes to to incoming data
